tests/unit/jira/test_queue.py

```diff
@@ -0,0 +1,196 @@
+from datetime import datetime, timedelta
+
+import pytest
+from pydantic import HttpUrl
+
+from jbi.queue import (
+    DeadLetterQueue,
+    FileBackend,
+    InvalidQueueDSNError,
+    QueueBackend,
+)
+
+
+@pytest.fixture
+def backend(tmp_path):
+    return FileBackend(tmp_path)
+
+
+@pytest.fixture
+def queue(tmp_path):
+    return DeadLetterQueue("file://" + str(tmp_path))
+
+
+@pytest.mark.parametrize(
+    "dsn", ["memory://", "http://www.example.com", HttpUrl("http://www.example.com")]
+)
+def test_invalid_queue_url(dsn):
+    with pytest.raises(InvalidQueueDSNError):
+        DeadLetterQueue(dsn)
+
+
+def test_ping(backend: QueueBackend):
+    assert backend.ping() is True
+
+
+def test_filebackend_ping_fails(caplog, tmp_path):
+    tmp_path.chmod(0o400)  # set to readonly
+    backend = FileBackend(tmp_path)
+    assert backend.ping() is False
+
+
+@pytest.mark.asyncio
+async def test_remove_last_item(backend: QueueBackend, queue_item_factory):
+    """When we remove the last item for a bug, we also remove it's key from the
+    backend"""
+
+    item = queue_item_factory()
+
+    await backend.put(item)
+    assert await backend.size() == 1
+
+    await backend.remove(item.payload.bug.id, item.identifier)
+    assert await backend.size() == 0
+
+
+@pytest.mark.asyncio
+async def test_clear(backend: QueueBackend, queue_item_factory):
+    item_1 = queue_item_factory(payload__bug__id=123)
+    item_2 = queue_item_factory(payload__bug__id=456)
+
+    await backend.put(item_1)
+    await backend.put(item_2)
+    assert await backend.size() == 2
+
+    await backend.clear()
+    assert await backend.size() == 0
+
+
+@pytest.mark.asyncio
+async def test_put_maintains_sorted_order(backend: QueueBackend, queue_item_factory):
+    now = datetime.now()
+    item_1 = queue_item_factory(payload__event__time=now + timedelta(minutes=1))
+    item_2 = queue_item_factory(payload__event__time=now + timedelta(minutes=2))
+    item_3 = queue_item_factory(payload__event__time=now + timedelta(minutes=3))
+    item_4 = queue_item_factory(payload__event__time=now + timedelta(minutes=4))
+
+    await backend.put(item_2)
+    await backend.put(item_1)
+    await backend.put(item_3)
+    await backend.put(item_4)
+
+    items = [item async for item in backend.get(item_1.payload.bug.id)]
+    assert list(items) == [item_1, item_2, item_3, item_4]
+
+
+@pytest.mark.asyncio
+async def test_get_all(backend: QueueBackend, queue_item_factory):
```

Comment:
Should we add a `test_get_all_failed` test where one file has invalid JSON in it?

Reply:
Good call. I added a few tests here [d6610f3](https://github.com/mozilla/jira-bugzilla-integration/pull/918/commits/d6610f316a30c40d3c424e2b270a47d34db2cfd5).

Should we catch at reraise these as our own QueueError or something?

Reply:
I'd say yes, just for the Slack notification to be explicit (ie. validation error can apply to incoming request or queue content now)

Reply:
👍🏻 [43a7a15](https://github.com/mozilla/jira-bugzilla-integration/pull/918/commits/43a7a158b6c80207fef65d6d0e66c6caa1234cdf)